File: src/test/java/pl/training/common/gestures/GestureImporter.java
package pl.training.common.gestures;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.appium.java_client.android.AndroidDriver;
import org.openqa.selenium.interactions.Pause;
import org.openqa.selenium.interactions.PointerInput;
import org.openqa.selenium.interactions.PointerInput.Origin;
import org.openqa.selenium.interactions.Sequence;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import static java.time.Duration.ofMillis;
import static org.openqa.selenium.interactions.PointerInput.Kind.TOUCH;
import static org.openqa.selenium.interactions.PointerInput.MouseButton.LEFT;

public class GestureImporter {

    private final ObjectMapper mapper;
    private final AndroidDriver driver;

    public GestureImporter(ObjectMapper mapper, AndroidDriver driver) {
        this.mapper = mapper;
        this.driver = driver;
    }

    public GestureImporter(AndroidDriver driver) {
        this.mapper = new ObjectMapper();
        this.driver = driver;
    }

    public List<Sequence> importGestures(String fileName) throws IOException {
        var gesture = mapper.readValue(new File(fileName), Gesture.class);
        var windowSize = driver.manage().window().getSize();

        var result = new ArrayList<Sequence>();
        var actions = gesture.getActions();
        for (int index = 0; index < actions.size(); index++) {
            var pointer = new PointerInput(TOUCH, "pointer" + index);
            var sequence = new Sequence(pointer, 1);
            var ticks = gesture.getActions().get(index).getTicks();
            for (Tick tick : ticks) {
                switch (tick.getType()) {
                    case "pointerMove":
                        sequence.addAction(pointer.createPointerMove(ofMillis(tick.getDuration()), Origin.viewport(),
                                (tick.getX().intValue() * windowSize.width) / 100,
                                (tick.getY().intValue() * windowSize.height) / 100));
                        if (tick.getDuration() == 0) {
                            sequence.addAction(new Pause(pointer, ofMillis(1_000)));
                        }
                        break;
                    case "pointerDown":
                        sequence.addAction(pointer.createPointerDown(LEFT.asArg()));
                        break;
                    case "pointerUp":
                        sequence.addAction(pointer.createPointerUp(LEFT.asArg()));
                        break;
                }
            }
            result.add(sequence);
        }
        return result;
    }

}

File: js/test/common/GestureImporter.js
const fs = require('fs').promises;
const { ActionSequence, Origin, WebDriver } = require('webdriverio');

class GestureImporter {
    constructor(driver) {
        this.driver = driver;
    }

    async import(fileName) {
        const data = await fs.readFile(fileName, 'utf-8');
        const gesture = JSON.parse(data);
        const windowSize = await this.driver.getWindowSize();
        const result = [];
        const actions = gesture.actions;

        for (let index = 0; index < actions.length; index++) {
            const pointer = {
                id:  "pointer" + index,
                type: 'pointer',
                parameters: { pointerType: 'touch' },
                actions: []
            };
            for (const tick of actions[index].ticks) {
                console.log(tick);
                switch (tick.type) {
                    case 'pointerMove':
                        pointer.actions.push({
                            type: 'pointerMove',
                            duration: tick.duration,
                            x: (tick.x * windowSize.width) / 100,
                            y: (tick.y * windowSize.height) / 100,
                            origin: "viewport"
                        });
                        if (tick.duration === 0) {
                            pointer.actions.push({type: 'pause', duration: 1000});
                        }
                        break;
                    case 'pointerDown':
                        pointer.actions.push({ type: 'pointerDown', button: 0 });
                        break;
                    case 'pointerUp':
                        pointer.actions.push({ type: 'pointerUp', button: 0 });
                        break;
                }
            }
            result.push(pointer);
        }
        return result;
    }

}

module.exports = GestureImporter;
